### PR TITLE
feat(search): adds support for minimalistic search requests

### PR DESCRIPTION
**Pull request by cs-marshall**
Includes definitions for a core set of term and match family queries.

### CODE REVIEW

File: search.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+// A simple term query search as an example:
+//
+//	req := NewSearchRequest()
+//	req.SetIndices("example_index")
+//	req.SetQuery(search.NewTermQuery("field", "basic")
+//	results, err := req.Do(context.Background(), client)
+type SearchRequest struct {
+    Query search.Query
+    Index []string
+    Size  *int
```
Comment (dylan-bourque) [inline search.go:25]: Why does this need to be a pointer?  Consumers will have to declare a separate variable and set this field to point to it, and that feels gross.

File: search.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+// NewSearchRequest instantiates an empty SearchRequest.
+// An empty SearchRequest will search across all indices and return the top 10 documents with the default [sorting].
+//
+// [sorting]: https://opensearch.org/docs/latest/opensearch/search/sort/
+func NewSearchRequest() *SearchRequest {
+    return &SearchRequest{}
+}
+
+// Source translates the SearchRequest into the shape expected by OpenSearch.
+func (r *SearchRequest) Source() (any, error) {
```
Comment (dylan-bourque) [inline search.go:38]: Is this so that `SearchRequest` satisfies the `search.Query` interface?  If so, call that out.

File: search.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        source["sort"] = sorts
+    }
+
+    return source, nil
+}
+
+// SetIndices sets the index list for the request.
+func (r *SearchRequest) SetIndices(indices ...string) *SearchRequest {
```
Comment (dylan-bourque) [inline search.go:66]: "Set" implies replacement, so maybe call this `AddIndices`.

File: search/boolean_query.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+// otherwise, it is not. The results of a filter query are generally cached to allow for a faster return.
+// Use the filter query to filter the results based on exact matches, ranges, dates, numbers, and so on.
+func (q *BoolQuery) Filter(queries ...Query) *BoolQuery {
+	q.filter = append(q.filter, queries...)
+	return q
+}
+
+// Source coverts the BoolQuery to the correct OpenSearch JSON
+func (q *BoolQuery) Source() (any, error) {
+	bq := make(map[string]interface{})
```
Comment (dylan-bourque) [inline search/boolean_query.go:63]: Use `any` for consistency

File: search.go (hunk 1 above)
Comment (cs-marshall) [reply search.go:25]: Consumers can pass in a normal int in the `SetSize` method, and then take to the pointer of that value. So they don't have to declare the separate variable.

I did it to control the default value. A normal int defaults to 0, so it'd be hard to tell the difference between a default and a value to ignore. I suppose I could have the constructor set it to -1 instead. That's probably a little cleaner...

File: search.go (hunk 2 above)
Comment (cs-marshall) [reply search.go:38]: It's not. It's just a convenient utility. It also allows consumers to call it and print out the json struct to debug their constructed search requests.

File: search/boolean_query.go (hunk 4 above)
Comment (cs-marshall) [reply search/boolean_query.go:63]: dang it, missed one!

File: search.go (hunk 1 above)
Comment (cs-marshall) [reply search.go]: Yes, I do it all the time on manual queries. Whenever I manually do an aggregation request, I don't care about the search results so I always set it to 0 to hide them. 

This is maybe not as applicable for an client wrapper. But it certainly is an optimization choice to reduce the data in flight.

File: search/boolean_query.go (hunk 4 above)
Comment (cs-marshall) [reply search/boolean_query.go]: I think I got them all now!

**Review by dylan-bourque — APPROVED** [review]

**Pull request merged**